src/lookup.cpp

```diff
@@ -65,33 +62,50 @@ SEXP find_promise_in_single_environment(const SEXP symbol, const SEXP rho) {
         return get_hash(hashcode, symbol, HASHTAB(rho));
     }
 
-    return R_UnboundValue;
+    return {lookup_status::SUCCESS, R_UnboundValue};
 }
 
-SEXP find_promise_in_environment(const SEXP symbol, const SEXP rho2) {
+lookup_result find_binding_in_environment(const SEXP symbol, const SEXP rho2) {
     SEXP rho = rho2;
 
     if (TYPEOF(rho) == NILSXP)
-        dyntrace_log_error("encountered NILSXP as environment");
+        return {lookup_status::FAIL_ENVIRONMENT_IS_NIL, R_UnboundValue};
 
     if (TYPEOF(rho) != ENVSXP)
-        dyntrace_log_error("second argument is not an environment");
+        return {lookup_status::FAIL_ARGUMENT_IS_NOT_AN_ENVIRONMENT, R_UnboundValue};
 
 #ifdef USE_GLOBAL_CACHE
-    while (rho != R_GlobalEnv && rho != R_EmptyEnv)
+    while (rho != R_GlobalEnv && rho != R_EmptyEnv) {
 #else
-    while (rho != R_EmptyEnv)
+    while (rho != R_EmptyEnv) {
 #endif
-    {
-        SEXP v = find_promise_in_single_environment(symbol, rho);
-        if (v != R_UnboundValue)
-            return v;
+        lookup_result r = find_binding_in_single_environment(symbol, rho);
+
+        if (r.status != lookup_status::SUCCESS)
+            return r;
+
+        if (r.value != R_UnboundValue)
+            return r;
+
         rho = ENCLOS(rho);
     }
 #ifdef USE_GLOBAL_CACHE
     if (rho == R_GlobalEnv)
         return find_promise_in_global_cache(symbol);
     else
 #endif
-        return R_UnboundValue;
+        return {lookup_status::SUCCESS, R_UnboundValue};
 }
+
+string lookup_status_to_string(lookup_status status) {
```

Comment:
Maybe we can rename this to `to_string` because C++ allows overloading functions on argument type.

Reply:
Maybe. What's the benefit?

Reply:
(This is a genuine question, I'm not trying to be difficult.)